Fixes for #2080 and #2081

wittibs opened the pull request:
Tested changes for #2080 and #2081 by publishing to npm and installing/running the package locally (specifically with --sass-binary-path CLI flag).

Fix for #2081 would be really nice. I work at a big corporation, and we have firewalls that prevent easy direct access to downloading sass binaries (which causes us some headaches). Being able to point to a compiled binary, using a *relative path*, would help us a lot.

Thank you!

xzyfer commented on lib/extensions.js:
```diff
@@ -154,15 +154,10 @@ function isSupportedEnvironment(platform, arch, abi) {
-
-  return flags[index + 1];
+function getArgument(name) {
+  var flag = process.argv.slice(2).find(a => a.indexOf(name) !== -1);
```
Please don't use ES6. We still support older versions of node

xzyfer commented on lib/binding.js:
```diff
@@ -3,18 +3,20 @@
   }
-
-  return require(ext.getBinaryPath());
+  if (!path.isAbsolute(binaryPath)) binaryPath = path.join('../../..', binaryPath);
```
Please keep to the existing code style.

```js
if (!path.isAbsolute(binaryPath)) {
  binaryPath = path.join('../../..', binaryPath);
}
```

xzyfer replied:
We can't assume *nix

```js
binaryPath = path.join('..', '..', '..', binaryPath);
```

xzyfer replied:
What is the rationale for going up exactly 3 directories?

wittibs replied:
I make the assumption that non-absolute binary paths are relative to the working project directory (and I specify this in the docs). Consider the following project structure:

```
project
|-index.html
|-index.js
|-my-node-binding
  |-binding.node
|-node_modules
  |-node-sass
    |-lib
      |-binding.js
|-package.json
|-package-lock.json
```

You can see binding.js is 3 directories deep relative to the working project directory (node-modules/node-sass/lib). The ``require()`` function loads modules/files relative to the calling file. Since binding.js ``require()``'s the binary, we have to go up 3 directories.

xzyfer replied:
This doesn't account for nested `node_modules` for instance.

```
project
|-index.html
|-index.js
|-my-node-binding
  |-binding.node
|-node_modules
  |-node-sass@4.5.2
    |-lib
      |-binding.js
  |-gulp-sass
    |-node_modules
      |-node-sass@3.7.1
        |-lib
          |-binding.js
|-package.json
|-package-lock.json
```

npm does not guarantee flat node_modules, and it's not possible to make such a guarantee when version conflicts happen. I think `process.cwd()` is probably a better fit.

wittibs commented:
Summary:
- removed ES6 ``find()`` function, replaced with ES5-compliant solution
- made code changes to adhere to coding style
- made suggested cross-platform update

Please review. Thank you!

xzyfer commented:
A few note. Removing support for the current `--foo bar` style is a breaking change. To get this change over the line we'd need to support both the old and new style. This can be done by asserting whether the matched value contains `=`.

>However that [won't work][1], as all custom CL arguments must have '--' prefixed.

I'm not sure I follow why this wont work given we're finding the index in the arg list the number of args beforehand shouldn't matter.

```
npm install node-sass -- --foo bar
```


[1]: https://stackoverflow.com/questions/11580961/sending-command-line-arguments-to-npm-script#answer-14404223

wittibs commented:
All custom CLI arguments must have '--' prefixed. So, this will fail:
```
npm install node-sass -- --foo bar
```
But this would work:
```
npm install node-sass -- --foo --bar
```
When you consider passing a binary path, we would need to strip the `--` from the beginning of the path.
```
npm install node-sass -- --sass-binary-path --./node_binary/binding.node
```
However, that's weird. I'd rather align with industry standard of arg=val, like so:
```
npm install node-sass -- --sass-binary-path=./node_binary/binding.node
```

wittibs commented:
Summary:
* modified `extensions.getBinaryPath()` to return an absolute path *from a relative path* if possible
  * it does this by attempting to `resolve()` the relative path against the current working directory (as you suggested earlier), but traverses up the directory tree trying to `resolve()` the path. This is how we get around the arbitrarily-nested `node_modules` directory problem
* undid changes to binding.js since they were no longer necessary

Please review. Thanks!

saper commented:
@wittibs is getting the absolute path really necessary?

wittibs commented:
@saper Yes - it is critical to ensuring relative binary paths will work for any level of depth of nested node_modules sub-directories.
For example: if I provide a binary path `./my-node-binding/binding.node` (relative to my project directory), I should expect the binary path to be resolved correctly regardless of node-sass directory depth. See sample file structure below (taken from @xzyfer 's comment from outdated code changes above):
```
project
|-index.html
|-index.js
|-my-node-binding
  |-binding.node
|-node_modules
  |-node-sass@4.5.2
    |-lib
      |-binding.js
  |-gulp-sass
    |-node_modules
      |-node-sass@3.7.1
        |-lib
          |-binding.js
|-package.json
|-package-lock.json
```

rdubigny commented:
Thx @wittibs for this work.

@zxyger can we merge this ASAP ?  We need this feature for our deployment pipeline to work :)

wittibs commented:
Hey @xzyfer just following up - have you had a chance to review this solution? If so, any feedback or thoughts?

Let me know if there's anything I can do to help expedite the PR. Thanks!

saper commented:
I am going to give it a try - how can Inreproduce the problem using Win Server 2016?

wittibs commented:
@saper try to install node-sass package with npmrc variable sass_binary_path set to a relative path. (#2081)

Try using any sass binary CLI argument during install or runtime (#2080)

saper commented:
Interesting problem when trying to use absolute paths: #2149

wittibs commented:
I'm considering closing this PR because:

- The fix for #2081 is non-deterministic, which is buggy in nature
- I don't think #2081 can be solved without some hackish fix. I think it should be the consuming application's responsibility to provide an absolute path for sass_binary_path.

Let me know what you all think. At the very least, I can:

- keep the fix for #2080 
- provide error message for when relative path is provided to sass_binary_path

Thoughts?

saper commented:
I've tried this today and it seems to work:

````> npm install node-sass --sass-binary-path=/tmp/binding.node

> node-sass@4.6.1 install /tmp/node/node_modules/node-sass
> node scripts/install.js

node-sass build Binary found at /tmp/binding.node

> node-sass@4.6.1 postinstall /tmp/node/node_modules/node-sass
> node scripts/build.js

Binary found at /tmp/binding.node
Testing binary
Binary is fine
````

I think something strange happened to npm since I believe the space separated form worked as well.

Can you try without `--` and with `=` separator?

wittibs commented:
I tried that and... it worked! Although I have no idea why .... I don't think I fully understand how CLI arguments get parsed & passed to node apps. :/

Well this has been a good learning experience for me. Thank you for your time.

jmcmullen commented:
@saper @wittibs Running into the same issue, but that fix has thrown me off completely. Can you tell me what I'm doing wrong?

```
$ node
> node-sass@4.6.1 install /tmp/lib/node-sass/linux-x64-51_binding.node
SyntaxError: Invalid or unexpected token
```

```
node-sass@4.6.1 install /tmp/lib/node-sass/linux-x64-51_binding.node
zsh: command not found: node-sass@4.6.1
```
